orc/src/main/java/com/netflix/iceberg/orc/TypeConversion.java

```diff
@@ -0,0 +1,172 @@
+      case FIXED:
+        return TypeDescription.createBinary();
+      case BINARY:
+        return TypeDescription.createBinary();
+      case DECIMAL: {
+        Types.DecimalType decimal = (Types.DecimalType) type;
+        return TypeDescription.createDecimal()
+            .withScale(decimal.scale())
+            .withPrecision(decimal.precision());
+      }
+      case STRUCT: {
+        TypeDescription struct = TypeDescription.createStruct();
+        for(Types.NestedField field: type.asStructType().fields()) {
+          columnIds.add(field.fieldId());
+          struct.addField(field.name(), toOrc(field.type(), columnIds));
+        }
+        return struct;
+      }
+      case LIST: {
+        Types.ListType list = (Types.ListType) type;
+        columnIds.add(list.elementId());
+        return TypeDescription.createList(toOrc(type.asListType().elementType(),
+            columnIds));
+      }
+      case MAP: {
+        Types.MapType map = (Types.MapType) type;
+        columnIds.add(map.keyId());
+        TypeDescription key = toOrc(map.keyType(), columnIds);
+        columnIds.add(map.valueId());
+        return TypeDescription.createMap(key,
+            toOrc(type.asMapType().valueType(), columnIds));
+      }
+      default:
+        throw new IllegalArgumentException("Unhandled type " + type.typeId());
+    }
+  }
+
+  /**
+   * Convert an ORC schema to an Iceberg schema.
+   * @param schema the ORC schema
+   * @param columnIds the column ids
+   * @return the Iceberg schema
+   */
+  public Schema fromOrc(TypeDescription schema, int[] columnIds) {
+    return new Schema(convertOrcToType(schema, columnIds).asStructType().fields());
+  }
+
+  Type convertOrcToType(TypeDescription schema, int[] columnIds) {
+    switch (schema.getCategory()) {
+      case BOOLEAN:
+        return Types.BooleanType.get();
+      case BYTE:
+      case SHORT:
+      case INT:
+        return Types.IntegerType.get();
+      case LONG:
+        return Types.LongType.get();
+      case FLOAT:
+        return Types.FloatType.get();
+      case DOUBLE:
+        return Types.DoubleType.get();
+      case STRING:
+      case CHAR:
+      case VARCHAR:
+        return Types.StringType.get();
+      case BINARY:
+        return Types.BinaryType.get();
+      case DATE:
+        return Types.DateType.get();
+      case TIMESTAMP:
+        return Types.TimestampType.withoutZone();
+      case DECIMAL:
+        return Types.DecimalType.of(schema.getPrecision(), schema.getScale());
+      case STRUCT: {
+        List<String> fieldNames = schema.getFieldNames();
+        List<TypeDescription> fieldTypes = schema.getChildren();
+        List<Types.NestedField> fields = new ArrayList<>(fieldNames.size());
+        for(int c=0; c < fieldNames.size(); ++c) {
+          String name = fieldNames.get(c);
+          TypeDescription type = fieldTypes.get(c);
+          fields.add(Types.NestedField.optional(columnIds[type.getId()], name,
```

Comment:
Does ORC not support a distinction between required and optional fields?

Because Iceberg assumes that the files in a given table are managed by Iceberg, I think we can probably work around this by adding non-null checks to the write path and assuming non-null on the read path. But I would still prefer to have a guarantee that the files won't contain null values.

Reply:
No, ORC doesn't support required field. The history of course is that Hive doesn't support required fields, so it wasn't that important. Once schema evolution was added it really doesn't make any sense to have required fields.

That said, the column statistics track whether there are any nulls in each column. So given a file footer you can tell whether there are any nulls in that column or not. You're right of course that it would be easy to check for null values on the Iceberg write path. On the read path, you always need to check for null because the column may not be present in the file.

Reply:
Schema evolution doesn't allow you to add a required column, so we shouldn't need to worry about missing required columns. I like that we can use the stats to make sure the file doesn't contain any nulls. Let's plan on doing that for the read path and throwing an exception.

Is it possible to add the metadata to ORC? We have a fairly compelling use case for it: when there are null values in a foreign key column, a SQL inner join will ignore the rows with null because joins use null-safe equality (null == null returns null). We want to ensure we don't accidentally lose rows this way because it is a subtle correctness issue and we can't expect users to know they should do an outer join everywhere. Instead, we want to ensure that the foreign key is *never* null.